File: backend/tests/unit/utils/dateUtils.test.ts
import { DateUtils } from '../../../src/utils/dateUtils';

describe('DateUtils', () => {
  describe('isLeapYear', () => {
    it('should return true for leap years', () => {
      expect(DateUtils.isLeapYear(2024)).toBe(true);
      expect(DateUtils.isLeapYear(2000)).toBe(true);
      expect(DateUtils.isLeapYear(2400)).toBe(true);
    });

    it('should return false for non-leap years', () => {
      expect(DateUtils.isLeapYear(2023)).toBe(false);
      expect(DateUtils.isLeapYear(1900)).toBe(false);
      expect(DateUtils.isLeapYear(2100)).toBe(false);
    });
  });

  describe('getDaysInMonth', () => {
    it('should return correct days for each month', () => {
      expect(DateUtils.getDaysInMonth(1, 2024)).toBe(31); // January
      expect(DateUtils.getDaysInMonth(2, 2024)).toBe(29); // February (leap year)
      expect(DateUtils.getDaysInMonth(2, 2023)).toBe(28); // February (non-leap)
      expect(DateUtils.getDaysInMonth(4, 2024)).toBe(30); // April
      expect(DateUtils.getDaysInMonth(12, 2024)).toBe(31); // December
    });
  });

  describe('isValidDate', () => {
    it('should validate correct dates', () => {
      expect(DateUtils.isValidDate(1, 15)).toBe(true);
      expect(DateUtils.isValidDate(2, 29)).toBe(true); // Feb 29 allowed
      expect(DateUtils.isValidDate(12, 31)).toBe(true);
    });

    it('should reject invalid dates', () => {
      expect(DateUtils.isValidDate(0, 15)).toBe(false); // Invalid month
      expect(DateUtils.isValidDate(13, 15)).toBe(false); // Invalid month
      expect(DateUtils.isValidDate(1, 0)).toBe(false); // Invalid day
      expect(DateUtils.isValidDate(2, 30)).toBe(false); // Feb 30
      expect(DateUtils.isValidDate(4, 31)).toBe(false); // April 31
    });
  });

  describe('getZodiacSign', () => {
    it('should return correct zodiac signs', () => {
      expect(DateUtils.getZodiacSign(1, 15)).toBe('capricorn');
      expect(DateUtils.getZodiacSign(3, 25)).toBe('aries');
      expect(DateUtils.getZodiacSign(7, 25)).toBe('leo');
      expect(DateUtils.getZodiacSign(12, 25)).toBe('capricorn');
    });

    it('should handle boundary dates correctly', () => {
      expect(DateUtils.getZodiacSign(1, 19)).toBe('capricorn');
      expect(DateUtils.getZodiacSign(1, 20)).toBe('aquarius');
      expect(DateUtils.getZodiacSign(12, 21)).toBe('sagittarius');
      expect(DateUtils.getZodiacSign(12, 22)).toBe('capricorn');
    });
  });

  describe('daysUntilBirthday', () => {
    it('should calculate days until birthday', () => {
      const today = new Date('2025-12-07');

      // Birthday in same month, after today
      expect(DateUtils.daysUntilBirthday(12, 15, today)).toBe(8);

      // Birthday already passed this year
      expect(DateUtils.daysUntilBirthday(1, 15, today)).toBeGreaterThan(30);

      // Birthday today
      expect(DateUtils.daysUntilBirthday(12, 7, today)).toBe(0);
    });
  });

  describe('calculateAge', () => {
    it('should calculate age correctly', () => {
      const today = new Date('2025-12-07');

      // Birthday already happened this year
      expect(DateUtils.calculateAge(2015, 1, 15, today)).toBe(10);

      // Birthday not yet this year
      expect(DateUtils.calculateAge(2015, 12, 25, today)).toBe(9);

      // Birthday today
      expect(DateUtils.calculateAge(2015, 12, 7, today)).toBe(10);
    });
  });
});

File: app/test/unit/utils/date_utils_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:birthday_app/core/utils/date_utils.dart';

void main() {
  group('DateHelper', () {
    group('isLeapYear', () {
      test('returns true for leap years', () {
        expect(DateHelper.isLeapYear(2024), isTrue);
        expect(DateHelper.isLeapYear(2000), isTrue);
        expect(DateHelper.isLeapYear(2400), isTrue);
      });

      test('returns false for non-leap years', () {
        expect(DateHelper.isLeapYear(2023), isFalse);
        expect(DateHelper.isLeapYear(1900), isFalse);
        expect(DateHelper.isLeapYear(2100), isFalse);
      });
    });

    group('getDaysInMonth', () {
      test('returns correct days for each month', () {
        expect(DateHelper.getDaysInMonth(1, 2024), equals(31)); // January
        expect(DateHelper.getDaysInMonth(2, 2024), equals(29)); // Feb (leap)
        expect(DateHelper.getDaysInMonth(2, 2023), equals(28)); // Feb (non-leap)
        expect(DateHelper.getDaysInMonth(4, 2024), equals(30)); // April
        expect(DateHelper.getDaysInMonth(12, 2024), equals(31)); // December
      });
    });

    group('daysUntilBirthday', () {
      test('calculates days until future birthday in same year', () {
        final testDate = DateTime(2025, 12, 7);
        final days = DateHelper.daysUntilBirthday(12, 15, from: testDate);
        expect(days, equals(8));
      });

      test('calculates days until birthday next year', () {
        final testDate = DateTime(2025, 12, 7);
        final days = DateHelper.daysUntilBirthday(1, 15, from: testDate);
        expect(days, greaterThan(30));
      });

      test('returns 0 for birthday today', () {
        final testDate = DateTime(2025, 12, 7);
        final days = DateHelper.daysUntilBirthday(12, 7, from: testDate);
        expect(days, equals(0));
      });
    });

    group('calculateAge', () {
      test('calculates age correctly when birthday passed', () {
        final testDate = DateTime(2025, 12, 7);
        final age = DateHelper.calculateAge(2015, 1, 15, from: testDate);
        expect(age, equals(10));
      });

      test('calculates age correctly when birthday not yet this year', () {
        final testDate = DateTime(2025, 12, 7);
        final age = DateHelper.calculateAge(2015, 12, 25, from: testDate);
        expect(age, equals(9));
      });

      test('calculates age correctly on birthday', () {
        final testDate = DateTime(2025, 12, 7);
        final age = DateHelper.calculateAge(2015, 12, 7, from: testDate);
        expect(age, equals(10));
      });
    });

    group('getZodiacSign', () {
      test('returns correct zodiac signs', () {
        expect(DateHelper.getZodiacSign(1, 15), equals('Capricorn'));
        expect(DateHelper.getZodiacSign(3, 25), equals('Aries'));
        expect(DateHelper.getZodiacSign(7, 25), equals('Leo'));
        expect(DateHelper.getZodiacSign(12, 25), equals('Capricorn'));
      });

      test('handles boundary dates correctly', () {
        expect(DateHelper.getZodiacSign(1, 19), equals('Capricorn'));
        expect(DateHelper.getZodiacSign(1, 20), equals('Aquarius'));
      });
    });

    group('formatDaysUntil', () {
      test('formats days correctly', () {
        expect(DateHelper.formatDaysUntil(0), equals('Today'));
        expect(DateHelper.formatDaysUntil(1), equals('Tomorrow'));
        expect(DateHelper.formatDaysUntil(5), equals('5 days'));
        expect(DateHelper.formatDaysUntil(10), equals('1 week'));
        expect(DateHelper.formatDaysUntil(20), equals('2 weeks'));
      });
    });
  });
}
